File: android/src/main/kotlin/com/pauloalvesjr/flutter_math_view/FlutterMathView.kt
package com.pauloalvesjr.flutter_math_view

import android.content.Context
import android.view.View
import android.view.ViewTreeObserver

import android.graphics.Color

import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

import com.agog.mathdisplay.MTMathView

class FlutterMathView(context: Context, messenger: BinaryMessenger, id: Int) : PlatformView, MethodCallHandler {
    private val mathView: MTMathView
    private val methodChannel: MethodChannel

    override fun getView(): View {
        return mathView
    }

    init {
        mathView = MTMathView(context)

        methodChannel = MethodChannel(messenger, "flutter_math_view$id")
        methodChannel.setMethodCallHandler(this)

        mathView.viewTreeObserver.addOnGlobalLayoutListener(object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                methodChannel.invokeMethod(
                        "size",
                        hashMapOf("height" to mathView.height, "width" to mathView.width)
                )
            }
        })

    }

    override fun onMethodCall(methodCall: MethodCall, result: MethodChannel.Result) {
        when (methodCall.method) {
            "setLatex" -> setLatex(methodCall, result)
            "setFontSize" -> setFontSize(methodCall, result)
            "setColor" -> setColor(methodCall, result)
            else -> result.notImplemented()
        }

    }

    private fun setLatex(methodCall: MethodCall, result: Result) {
        val latex = methodCall.arguments as String
        mathView.latex = latex
        result.success(null)
    }

    private fun setFontSize(methodCall: MethodCall, result: Result) {
        val fontSize = methodCall.arguments as Double
        mathView.fontSize = MTMathView.convertDpToPixel(fontSize.toFloat())
        result.success(null)
    }

    private fun setColor(methodCall: MethodCall, result: Result) {
        val colorString = methodCall.arguments as String
        mathView.textColor = Color.parseColor(colorString)
        result.success(null)
    }

    override fun dispose() {
        // TODO dispose actions if needed
    }
}

File: ios/Classes/FlutterMathView.swift
//
//  FlutterMathView.swift
//  flutter_math_view
//
//  Created by Paulo Alves on 27/08/20.
//

import Flutter
import UIKit

import iosMath

public class FlutterMathView : NSObject, FlutterPlatformView {
    let frame: CGRect
    let viewId: Int64
    let channel: FlutterMethodChannel
    let mathView: MTMathUILabel
    var isInitialized: Bool = false

    init(_ frame: CGRect, viewId: Int64, channel: FlutterMethodChannel, args: Any?) {
        self.frame = frame
        self.viewId = viewId
        self.channel = channel
        self.mathView = MTMathUILabel(frame: frame)
        
        super.init()
        
        channel.setMethodCallHandler({ (methodCall: FlutterMethodCall, result: FlutterResult) -> Void in
            switch methodCall.method {
            case "setLatex":
                self.setLatex(methodCall: methodCall, result: result)
            case "setFontSize":
                self.setFontSize(methodCall: methodCall, result: result)
            case "setColor":
                self.setColor(methodCall: methodCall, result: result)
            default:
                result(FlutterMethodNotImplemented)
            }
        })
    }
    
    private func setLatex(methodCall: FlutterMethodCall, result: FlutterResult) {
        let latex = methodCall.arguments as? String
        self.mathView.latex = latex
    }
    
    private func setFontSize(methodCall: FlutterMethodCall, result: FlutterResult) {
        let fontSize = methodCall.arguments as? Double
        self.mathView.fontSize = CGFloat(fontSize!)
    }
    
    private func setColor(methodCall: FlutterMethodCall, result: FlutterResult) {
        let colorString = methodCall.arguments as! String
        self.mathView.textColor = UIColor(hex: colorString)!
    }
    
    public func view() -> UIView {
        let height = self.mathView.intrinsicContentSize.height
        let width = self.mathView.intrinsicContentSize.width
        
        if(!isInitialized && height != 0 && width != 0) {
            isInitialized = true
            channel.invokeMethod("size", arguments: [
                "height": self.mathView.intrinsicContentSize.height,
                "width": self.mathView.intrinsicContentSize.width
            ])
        }
        
        return self.mathView
    }
    
    
}
